ls1intum/bamboo-server-notification-plugin
# Checkstyle and PMD report parsing

### Motivation and Context
This PR extends the parsing functionality for static code analysis reports.

### Description

- Add XOM for parsing
- Add PMD parsing
- Add Checkstyle parsing
- Switch to parsing of the more verbose spotbugsXml, as it contains a proper source path. Opened an issue to add this information to the xdoc report but this might take a while.
- Rename classname to filePath
- Rename type to rule
- Add PR template

### Steps for Testing


1. Build and deploy the notification locally
2. Start Artemis in debug mode and create a programming exercise with static code analysis enabled
3. Set break point in `ResultResource.notifyNewResult`
4. Alter pom.xml in the test repository to include the Checkstyle and Pmd maven plugin (example can be found https://github.com/kloessst/BambooTest/blob/master/pom.xml). Just copy those dependencies.
5. Alter the static code analysis task to match:
![scaTask](https://user-images.githubusercontent.com/16407766/91561927-a55f0980-e93c-11ea-9d9b-492dcdb1d927.PNG)
6. Alter the artifacts to match:
![artifacts](https://user-images.githubusercontent.com/16407766/91561986-ba3b9d00-e93c-11ea-832f-8271a6c2be9e.PNG)
7. Participate in the programming exercise and commit bad code
8. Inspect the request body when the break point triggers
--> Static Code Analysis reports should be part of the request body and should match the format defined in the notification plugin readme.
